Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -332,15 +332,16 @@ public class ProgramBuilder {
     /// Required when copying instructions between program.
     ///
     private var varMaps = [VariableMap<Variable>]()
-    private var typeMaps = [VariableMap<[Int: Type]>]()
+    /// Inverted ProgramTypes structure for easier adopting of runtimeTypes
+    private var typeMaps = [[Int: [(Variable, Type)]]]()
```

Comment:
Would it be more efficient here to use a simple Array as outer data structure instead of a Map since the integers will be mostly contiguous and starting at 0?

Reply:
There are only 40-50% of instructions covered (we don't collect literals,...)

Reply:
But wouldn't an array be more storage efficient even with 25% or less? I'd assume the Map would be overcommitted by roughly a factor of two, and, depending on how it is implemented, would use element structs for its implementation which are probably at least twice the size of an Int. Performance-wise the array should easily win in any case.

Reply:
TBH I've never done any experiment to get the constant factor overhead.

But in that case we have a bit more complicated implementation and maybe this deserves own struct too?

Reply:
Ah hm, I partly also suggested this because I thought it would make the implementation easier :'D
So I think it should be fine to make this `[[(Variable, Type)]]`, then initialize it fully in `beginAdoption`: `var typeMap = Array(repeating: [], count: program.size)`. That way, no nil handling is required during construction and use of the type maps. This should be pretty much the fastest in terms of execution speed also (which I mostly care about).
Regarding memory consumption, I think at least once the ProgramType information also includes the information from the AbstractInterpreter, we probably will have type information for most lines, so preallocating the array in full shouldn't cause much overhead. I also don't really think that using `nil` values for missing entries would bring much gain, since the size of that `nil` should also be the size of an empty array (the element size of the array must be the size of the largest element) (that assumes that empty arrays don't preallocate their backing memory on the heap though).

Reply:
nice this is also implementation-wise easy.
Only one thing. Type information will not cover more instructions. Because there will be 2 separate structures. One for runtime types (for future adoption) and second for types infered by AbstractInterpreter (for queries)

Reply:
Yes I think that's correct (if I understand correctly). I think there will the the `typeMap`, which is constructed once during `beginAdoption` for *another* (immutable) program and then only queried (until `endAdoption`) but not updated. Then there will be a `var types: ProgramTypes` member of the ProgramBuilder which will hold the types for the program *under construction* and which will be updated from adopted types and from the interpreter. Then, during `finalize()`, that types data structure will be attached to the produced Program. Is that also how you imagine it will work?